libraries/GCS_MAVLink/GCS_MAVFTP.cpp

```diff
@@ -0,0 +1,713 @@
+void GCS_MAVLINK::_ftp_burst_send()
+{
+    // Anything to stream?
+    if (!_session_info.stream_download) {
+        return;
+    }
+    FTPPayloadHeader *stream_payload = reinterpret_cast<FTPPayloadHeader *>(&_ftp_response.payload[0]);
+
+    // Send stream packets until buffer is full
+
+    // Skip send if not enough room
+    if (!HAVE_PAYLOAD_SPACE(chan, FILE_TRANSFER_PROTOCOL)) {
+        return;
+    }
+
+    bool more_data;
+
+    do {
+        more_data = false;
+        FTPErrorCode error_code = kErrNone;
+
+        stream_payload->seq_number = _session_info.stream_seq_number;
+        stream_payload->session = 0;
+        stream_payload->opcode = kRspAck;
+        stream_payload->req_opcode = kCmdBurstReadFile;
+        stream_payload->offset = _session_info.stream_offset;
+        _session_info.stream_seq_number++;
+
+
+        // We have to test seek past EOF ourselves, lseek will allow seek past EOF
+        if (_session_info.stream_offset >= _session_info.file_size) {
+            error_code = kErrEOF;
+        }
+
+        if (error_code == kErrNone) {
+            if (lseek(_session_info.fd, stream_payload->offset, SEEK_SET) < 0) {
+                _ftp_errno = errno;
+                error_code = kErrFailErrno;
+            }
+        }
+
+        if (error_code == kErrNone) {
+            int bytes_read = ::read(_session_info.fd, &stream_payload->data[0], MAX_FTP_DATA_LENGTH);
+
+            if (bytes_read < 0) {
+                // Negative return indicates error other than eof
+                _ftp_errno = errno;
+                error_code = kErrFailErrno;
+            }
+            else {
+                stream_payload->size = bytes_read;
+                _session_info.stream_offset += bytes_read;
+                _session_info.stream_chunk_transmitted += bytes_read;
+            }
+        }
+
+        if (error_code != kErrNone) {
+            stream_payload->opcode = kRspNak;
+            stream_payload->size = 1;
+            uint8_t *pData = &stream_payload->data[0];
+            *pData = error_code; // Straight reference to data[0] is causing bogus gcc array subscript error
+
+            if (error_code == kErrFailErrno) {
+                _ftp_errno = errno;
+                stream_payload->size = 2;
+                stream_payload->data[1] = _ftp_errno;
+            }
+
+            _session_info.stream_download = false;
+
+        }
+        else {
+            more_data = true;
+            stream_payload->burst_complete = false;
+        }
+        _ftp_response.target_system = _session_info.stream_target_system_id;
+        mavlink_msg_file_transfer_protocol_send_struct(chan, &_ftp_response);
+
+        if (more_data && !HAVE_PAYLOAD_SPACE(chan, FILE_TRANSFER_PROTOCOL)) {
+            more_data = false;
+
+            /* perform transfers in 35K chunks - this is determined empirical */
+            if (_session_info.stream_chunk_transmitted > 35000) {
+                stream_payload->burst_complete = true;
```

Comment:
this is never sent. as the packet is sent at line #698 so there is no indication that this is the last transmission.
my implementation is hitting a timeout retry loop because of it.